### PR TITLE
Windows Auto Updater

### PR DESCRIPTION
**Pull request by patrickdet**
Most of this has just been extracted from https://github.com/atom/atom. The Api of the current Mac autoupdater is kept. Feedback would be greatly appreciated.

### CODE REVIEW

**Comment by patrickdet** [comment]
tracked at https://github.com/atom/electron/issues/939

File: atom/browser/api/lib/auto-updater/auto-updater-win.coffee [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{EventEmitter} = require 'events'
+SquirrelUpdate = require './auto-updater/squirrel-update-win'
+
+class AutoUpdater
+  @__proto__ = EventEmitter.prototype
+
+  quitAndInstall: ->
+    # TODO
+
+  setFeedUrl: (@updateUrl) ->
```
Comment (paulcbetts) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]: Can you just set it explicitly? This is clever but a little _too_ clever :)

File: atom/browser/api/lib/auto-updater/squirrel-update-win.coffee [hunk 2]
```diff
@@ -0,0 +1,42 @@
+ChildProcess = require 'child_process'
+fs           = require 'fs'
+path         = require 'path'
+
+appFolder             = path.resolve(process.execPath, '..') # i.e. my-app/app-0.1.13/
```
Comment (paulcbetts) [inline atom/browser/api/lib/auto-updater/squirrel-update-win.coffee]: `path.dirname(process.execPath)`

**Comment by paulcbetts** [comment]
This looks pretty good so far - one thing we could think about is how we can implement squirrel event hooks and help people to create shortcuts (i.e. all the jazz that happens in https://github.com/atom/atom/blob/master/src/browser/squirrel-update.coffee#L230)

File: atom/browser/api/lib/auto-updater.coffee [hunk 3]
```diff
@@ -1,4 +1,9 @@
-autoUpdater = process.atomBinding('auto_updater').autoUpdater
+switch process.platform
+  when 'darwin'
+    autoUpdater = process.atomBinding('auto_updater').autoUpdater
+  when 'win32'
+    autoUpdater = require('./auto-updater-win')
+
 EventEmitter = require('events').EventEmitter
 
 autoUpdater.__proto__ = EventEmitter.prototype
```
Comment (kevinsawicki) [inline atom/browser/api/lib/auto-updater.coffee]: Won't this blow up on Linux since `autoUpdater` will be `null` in this case?

Comment (patrickdet) [inline atom/browser/api/lib/auto-updater.coffee]: very true. will change it. thx

File: atom/browser/api/lib/auto-updater/auto-updater-win.coffee (hunk 1 above)
Comment (patrickdet) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]: you mean the `setFeedUrl: (@updateUrl)`? sure, will do

**Comment by patrickdet** [comment]
@paulcbetts yeah i will come up with a default handler for the squirrel events. something that is extensible

File: atom/browser/api/lib/auto-updater/auto-updater-win.coffee [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{EventEmitter} = require 'events'
+SquirrelUpdate = require './auto-updater/squirrel-update-win'
+
+class AutoUpdater
+  @__proto__ = EventEmitter.prototype
```
Comment (steelbrain) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]: why not do `class AutoUpdate extends EventEmitter`?

Comment (patrickdet) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]: good point!

**Comment by patrickdet** [comment]
@paulcbetts i have cleaned up the current pull request. still missing: squirrel events default handler.

Are you looking for something like this? 

main.coffee
```Coffeescript
app           = require('app')
BrowserWindow = require('browser-window')
crashReporter = require('crash-reporter')
autoUpdate    = require('auto-update')
mainWindow    = null


crashReporter.start()

return if autoUpdater.squirrelEventsDefaultHandler?()

autoUpdater.setFeedUrl('http://mycompany.com/myapp/latest');

... etc
```

the downside is that this default handler has to be called explicitly. i could of course run it by default in the  `auto-updater.coffee` but then the user has no way of disabling it.

**Comment by jessep** [comment]
Any news on this? I'm currently reading through the atom/electron source to try understanding how to do all of the update logic properly. If this is going to become the state of the art soon, would be good to know :)

**Comment by paulcbetts** [comment]
@patrickdet I'll have a look at this soon, sorry for spacing on it

**Comment by pracucci** [comment]
As far as I can see, when you run `Update.exe --download` it will check `RELEASES` and if an update is available will download the `.nupkg` file and finally emit `update-available`.

Then you run `Update.exe --update` that will download again the `.nupkg` file (according to my tests), will run the new application version with `--squirrel-install` and finally emit `update-downloaded`.

I can see two not-optimal behaviours, but I'm not sure yet if we can fix it:
1. The `.nupkg` file (at least 40M) will be downloaded 2 times
2. `update-downloaded` is emitted once the update has been downloaded **and applied** (and it also run the application, so you see the new application window flash for few seconds)

**Comment by jessep** [comment]
@pracucci If the .nupkg is already downloaded, maybe the windows update script .update should call "Update.exe --install" rather than "Update.exe --update"? Seems like that would avoid the second download.

The [docs](https://github.com/Squirrel/Squirrel.Windows/blob/0d1250aa6f0c25fe22e92add78af327d1277d97d/specs/Installer.md) seem to imply that it would just return metadata, but it isn't totally clear. If it is actually downloading the file, seems like --install would be the way to go. Did you try that? 

@paulcbetts @patrickdet I'm currently working on building an electron app for my product (workflowy.com) and I'd like to use this if possible. Is there anything I can do to help push this forward? Or should I just do a manual build from patrick's repo?

**Comment by paulcbetts** [comment]
`Update.exe --install` is an internal command, you definitely want `update.exe --update` to finally apply the update. `update.exe --download` might suck and be broken, we need to fix it up. 

> The .nupkg file (usually a large file) will be downloaded 2 times

This is weird, we should just fix that.

**Comment by pracucci** [comment]
@jessep, @paulcbetts is right. `Update.exe --install` is an internal command and can't be called to install an update after you call `Update.exe --download`. I spent some time doing experiments and I still didn't find a solution to avoid the double download (except patching Squirrel for Windows, that was out of my scope). Any help from @paulcbetts here is very appreciated.

**Comment by paulcbetts** [comment]
> The docs seem to imply that it would just return metadata, but it isn't totally clear. 

So the problem is, that in order to return the release notes, we have to actually download the file - this is an unfortunate aspect to Squirrel's design, but it makes `RELEASES` super small and super easy to parse / edit. Tradeoffs!

**Comment by pracucci** [comment]
@paulcbetts Is downloading `RELEASES` from the upcoming `auto-updater` module version for Windows, a viable solution or not? The workflow would be: manually check `RELEASES`, then `Update.exe --update`.

**Comment by petrbrzek** [comment]
Any update on this?

**Comment by jessep** [comment]
@paulcbetts It sound like you think the current pull request is adequate? Any reason not to merge this pull request in that case?

**Comment by jessep** [comment]
@paulcbetts nevermind, i just saw that there's the outstanding question about [squirrel events default handler](https://github.com/atom/electron/pull/1984#issuecomment-114137509) anyway, i'm eager to use this :/

**Comment by etiktin** [comment]
@paulcbetts, could you answer the [squirrel events default handler](https://github.com/atom/electron/pull/1984#issuecomment-114137509) question, so we can move forward with this?

**Comment by patrickdet** [comment]
Whenever the question is answered I will gladly update my PR. I also want the auto-updater for my own projects ;)

File: atom/browser/api/lib/auto-updater/auto-updater-win.coffee [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{EventEmitter} = require 'events'
+SquirrelUpdate = require './auto-updater/squirrel-update-win'
+app            = require('app')
+url            = require('url')
```
Comment (lipis) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]:
```coffeescript
app            = require 'app'
url            = require 'url'
```

File: atom/browser/api/lib/auto-updater/squirrel-update-win.coffee [hunk 6]
```diff
@@ -0,0 +1,57 @@
+ChildProcess = require 'child_process'
+fs           = require 'fs'
+path         = require 'path'
+
+appFolder             = path.dirname(process.execPath) # i.e. my-app/app-0.1.13/
+rootApplicationFolder = path.resolve(appFolder, '..')  # i.e. my-app/
+updateDotExe          = path.join(rootApplicationFolder, 'Update.exe')
+exeName               = path.basename(process.execPath)
```
Comment (lipis) [inline atom/browser/api/lib/auto-updater/squirrel-update-win.coffee]: What's the policy with parentheses?

```coffeescript
appFolder             = path.dirname process.execPath # i.e. my-app/app-0.1.13/
rootApplicationFolder = path.resolve appFolder, '..'  # i.e. my-app/
updateDotExe          = path.join rootApplicationFolder, 'Update.exe'
exeName               = path.basename process.execPath
```

**Comment by etiktin** [comment]
@patrickdet, in regards to [squirrel events default handler](https://github.com/atom/electron/pull/1984#issuecomment-114137509), I think it's fine to call the handler explicitly and pass to it what ever it needs. That's pretty much the way it's implemented in [atom](https://github.com/atom/atom/blob/master/src/browser/main.coffee#L51-L55) (on start it calls `handleStartupEventWithSquirrel` which sets up everything).

I think we should go a head and finish it. If we are not sure about the quality of the API, we can mark it as "experimental" (AKA it might break in future versions) in the docs and let people start playing with it so we can get real feedback.

**Comment by paulcbetts** [comment]
Sorry for totally spacing on this, lemme read through a few things today

**Comment by petrbrzek** [comment]
@paulcbetts How does it look? :)

**Comment by michaelzoidl** [comment]
+1

**Comment by sevengo8378** [comment]
+1

File: atom/browser/api/lib/auto-updater/auto-updater-win.coffee [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      app.quit()
+
+  setFeedUrl: (updateUrl) ->
+    # set feed URL only when it hasn't been set before
+    unless @updateUrl
+      # See https://github.com/Squirrel/Squirrel.Windows/issues/132
+      # This way the Mac and Windows Update URL can be the same, even when
+      # the Mac version is sending additional data in the query string.
+      parsedUrl = url.parse(updateUrl)
+      delete parsedUrl.search
```
Comment (paulcbetts) [inline atom/browser/api/lib/auto-updater/auto-updater-win.coffee]: You actually don't need to do this anymore in the latest version of Squirrel

**Comment by paulcbetts** [comment]
@patrickdet Let's get this in! :shipit:

**Comment by zcbenz** [comment]
The new files should be added to `filenames.gypi`.

**Comment by patrickdet** [comment]
Cool! I will fix the cosmetic issues tomorrow and then it should be good to
go.

**Comment by jessep** [comment]
Huzzaaahhhh!!!!

**Comment by hankbao** [comment]
@patrickdet any update here? Thanks.

**Comment by patrickdet** [comment]
@paulcbetts @zcbenz ready when you are :shipit:

**Comment by zcbenz** [comment]
Awesome! :+1:

**Comment by michaelzoidl** [comment]
Should we update the docs? the auto-updater doc still says windoof is not supported..
https://github.com/atom/electron/blob/master/docs/api/auto-updater.md

**Comment by lipis** [comment]
+1 for the docs!

**Comment by lipis** [comment]
@michaelzoidl Maybe you should create a new issue to keep track of it.

**Comment by michaelzoidl** [comment]
@lipis done.

**Pull request merged**